### PR TITLE
[21121] IPayloadPool refactor 

### PR DESCRIPTION
**Pull request by cferreiragonz**
## Description

This PR adds two main changes:

1. It moves `payload_owner_` from `CacheChange_t` to `SerializedPayload_t`
2. Refactors `IPayloadPool` class and all derived classes to only manage `SerializedPayload_t`, instead of `CacheChange_t`
3. `get_payload()` methods refactor
4. `SerializedPayload_t` copies are forbidden

In this way, the `payload_owner_` is no longer related to a cache change, but to the payload it refers.

## Contributor Checklist

- [X] Commit messages follow the project guidelines. 
- [X] The code follows the style guidelines of this project. 
- [X] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- [X] Any new/modified methods have been properly documented using Doxygen. 
- _N/A_ Any new configuration API has an equivalent XML API (with the corresponding XSD extension) 
- ❌ Changes are backport compatible: they do **NOT** break ABI nor change library core behavior. 
- ❌ Changes are API compatible. 
- [x] New feature has been added to the `versions.md` file (if applicable).
- [X] New feature has been documented/Current behavior is correctly described in the documentation.
    - Related documentation PR: https://github.com/eProsima/Fast-DDS-docs/pull/800
 - _N/A_ Applicable backports have been included in the description.

## Reviewer Checklist

- [x] The PR has a milestone assigned.
- [x] The title and description correctly express the PR's purpose.
- [x] Check contributor checklist is correct.
- [x] Check CI results: changes do not issue any warning.
- [x] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

**Comment by cferreiragonz** [comment]
Since the `payload_owner_` now belongs to the `SerializedPayload_t`, I believe the [`get_payload()`](https://github.com/eProsima/Fast-DDS/blob/1f1d426775a21af4ddd178c5e323e3d47c638e45/include/fastdds/rtps/history/IPayloadPool.h#L94) method receiving a source `SerializedPayload_t&`, an `IPayloadPool*` and a destination `SerializedPayload_t&` could be simplified removing the `IPayloadPool*` parameter, as it is already contained in the source payload. I let the reviewer the decision of keeping it or not.

File: include/fastdds/rtps/history/IPayloadPool.h [hunk 1]
```diff
@@ -49,78 +47,68 @@ class IPayloadPool
...
-     *
-     * @warning @c data fields can only be changed when @c data_owner is @c nullptr. If a value different from
-     * @c nullptr is received all fields in @c data should be left unchanged.
+     * @warning @c data_owner contained in @c data can only be changed from @c nullptr to @c this. If a value
+     * different from @c nullptr is received it should be left unchanged.
      *
-     * @pre
-     *     @li Field @c cache_change.writerGUID is not @c unknown
-     *     @li Field @c cache_change.sequenceNumber is not @c unknown
+     * @warning @c data fields can only be changed when @c data_owner contained in @c data is @c nullptr. If a
```
Comment (MiguelCompany) [inline include/fastdds/rtps/history/IPayloadPool.h:83]:
```suggestion
     * @warning @c data fields can only be changed when @c payload_owner contained in @c data is @c nullptr. If a
```

File: src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp [hunk 2]
```diff
@@ -131,9 +131,7 @@ struct SampleLoanManager
         // Increment references of input payload
         CacheChange_t tmp;
         tmp.copy_not_memcpy(change);
-        item->owner = change->payload_owner();
-        change->payload_owner()->get_payload(change->serializedPayload, item->owner, tmp);
-        item->owner = tmp.payload_owner();
+        change->payload_owner()->get_payload(change->serializedPayload, tmp.serializedPayload);
         item->payload = tmp.serializedPayload;
         tmp.payload_owner(nullptr);
         tmp.serializedPayload.data = nullptr;
```
Comment (MiguelCompany) [inline src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp:137]: We can now avoid using a temporary `CacheChange_t` here, and use `change->payload_owner()->get_payload(change->serializedPayload, item->payload);`

File: src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp [hunk 3]
```diff
@@ -165,11 +163,11 @@ struct SampleLoanManager
...
         {
             CacheChange_t tmp;
-            tmp.payload_owner(item->owner);
             tmp.serializedPayload = item->payload;
-            item->owner->release_payload(tmp);
+            tmp.payload_owner(item->payload_owner());
+            item->payload_owner()->release_payload(tmp.serializedPayload);
             item->payload.data = nullptr;
-            item->owner = nullptr;
+            item->payload_owner(nullptr);
```
Comment (MiguelCompany) [inline src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp:170]: Similar thing here, could use `item->payload.payload_owner->release_payload(item->payload);`

File: src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp [hunk 4]
```diff
@@ -184,12 +182,13 @@ struct SampleLoanManager
         void* sample = nullptr;
         SampleIdentity identity;
         SerializedPayload_t payload;
-        IPayloadPool* owner = nullptr;
         uint32_t num_refs = 0;
 
         ~OutstandingLoanItem()
         {
+            // Avoid releasing payload and freeing data
```
Comment (MiguelCompany) [inline src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp:189]: We could just do:
```suggestion
            // Avoid releasing payload and freeing data
            payload.payload_owner = nullptr;
```

File: src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp [hunk 5]
```diff
@@ -208,6 +207,22 @@ struct SampleLoanManager
...
+        {
+            return payload.payload_owner();
+        }
+
+        void payload_owner(
+                IPayloadPool* owner)
+        {
+            payload.payload_owner(owner);
+        }
+
```
Comment (MiguelCompany) [inline src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp:225]: I don't think these are necessary

File: src/cpp/rtps/DataSharing/ReaderPool.hpp [hunk 6]
```diff
@@ -44,41 +44,57 @@ class ReaderPool : public DataSharingPayloadPool
...
+        return false;
+    }
 
-        // If owner is not this, then it must be an intraprocess datasharing writer
-        assert(nullptr != dynamic_cast<DataSharingPayloadPool*>(data_owner));
-        PayloadNode* payload = PayloadNode::get_from_data(data.data);
+    /**
+     * Prepares and fills the change in the datasharing protocol.
+     *
+     * If the payload is not owned by @c data_owner, it is assumed to be an intraprocess writer and the
```
Comment (MiguelCompany) [inline src/cpp/rtps/DataSharing/ReaderPool.hpp:71]:
```suggestion
     * If the payload is not owned by this pool, it is assumed to be an intraprocess writer and the
```

File: src/cpp/rtps/common/PayloadInfo_t.hpp [hunk 7]
```diff
@@ -65,14 +65,14 @@ struct PayloadInfo_t
...
 
-        change.payload_owner(payload_owner);
-        payload_owner = nullptr;
-
         change.serializedPayload = payload;
         payload.data = nullptr;
         payload.length = 0;
         payload.pos = 0;
         payload.max_size = 0;
+
```
Comment (MiguelCompany) [inline src/cpp/rtps/common/PayloadInfo_t.hpp:73]: I think we could get rid of `PayloadInfo_t` and use `SerializedPayload_t` directly.

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: include/fastdds/rtps/common/SerializedPayload.h [hunk 8]
```diff
@@ -91,6 +95,12 @@ struct FASTDDS_EXPORTED_API SerializedPayload_t
 
     ~SerializedPayload_t()
     {
+        if (payload_owner != nullptr)
+        {
+            payload_owner->release_payload(*this);
+        }
+        assert(payload_owner == nullptr);
+
```
Comment (MiguelCompany) [inline include/fastdds/rtps/common/SerializedPayload.h:103]: This code should be moved inside the empty() method

File: src/cpp/fastdds/publisher/DataWriterImpl.cpp [hunk 9]
```diff
@@ -995,7 +995,9 @@ ReturnCode_t DataWriterImpl::perform_create_new_change(
     CacheChange_t* ch = writer_->new_change(change_kind, handle);
     if (ch != nullptr)
     {
-        payload.move_into_change(*ch);
+        ch->serializedPayload = payload;
+        payload.data = nullptr;
+        payload.payload_owner = nullptr;
```
Comment (MiguelCompany) [inline src/cpp/fastdds/publisher/DataWriterImpl.cpp:1000]: Consider using `std::swap(payload, ch->serializedPayload);`

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

**Review by MiguelCompany — DISMISSED** [review]
LGTM with green CI

File: src/cpp/rtps/reader/StatefulReader.cpp [hunk 10]
```diff
@@ -641,7 +641,15 @@ bool StatefulReader::processDataMsg(
                 }
                 datasharing_pool->get_datasharing_change(change->serializedPayload, *change_to_add);
             }
-            else if (!payload_pool_->get_payload(change->serializedPayload, change_to_add->serializedPayload))
+            else if (payload_pool_->get_payload(change->serializedPayload, change_to_add->serializedPayload))
+            {
+                if (change->serializedPayload.payload_owner == nullptr)
+                {
+                    change_to_add->serializedPayload.payload_owner->get_payload(change_to_add->serializedPayload,
+                            change->serializedPayload);
```
Comment (MiguelCompany) [inline src/cpp/rtps/reader/StatefulReader.cpp:649]: We can try the following:

```suggestion
                    payload_pool_->get_payload(change_to_add->serializedPayload, change->serializedPayload);
```

File: src/cpp/rtps/reader/StatelessReader.cpp [hunk 11]
```diff
@@ -635,7 +635,15 @@ bool StatelessReader::processDataMsg(
                 }
                 datasharing_pool->get_datasharing_change(change->serializedPayload, *change_to_add);
             }
-            else if (!payload_pool_->get_payload(change->serializedPayload, change_to_add->serializedPayload))
+            else if (payload_pool_->get_payload(change->serializedPayload, change_to_add->serializedPayload))
+            {
+                if (change->serializedPayload.payload_owner == nullptr)
+                {
+                    change_to_add->serializedPayload.payload_owner->get_payload(change_to_add->serializedPayload,
```
Comment (MiguelCompany) [inline src/cpp/rtps/reader/StatelessReader.cpp:642]: Same here

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: include/fastdds/rtps/common/SerializedPayload.h [hunk 12]
```diff
@@ -83,6 +83,13 @@ struct FASTDDS_EXPORTED_API SerializedPayload_t
     {
     }
 
+    //!Copy constructor
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply include/fastdds/rtps/common/SerializedPayload.h:86]: Also move the implementation of the methods to a source file, please.

**Review by MiguelCompany — DISMISSED** [review]
LGTM with green CI
🏅

**Comment by MiguelCompany** [comment]
@cferreiragonz I have rebased this on top of #4875. When it gets merged, please re-rebase and update `versions.md` on this PR.

File: test/unittest/rtps/reader/CMakeLists.txt [hunk 13]
```diff
@@ -204,6 +204,7 @@ set(STATEFUL_READER_TESTS_SOURCE StatefulReaderTests.cpp
     ${PROJECT_SOURCE_DIR}/src/cpp/rtps/builtin/liveliness/WLP.cpp
     ${PROJECT_SOURCE_DIR}/src/cpp/rtps/builtin/liveliness/WLPListener.cpp
     ${PROJECT_SOURCE_DIR}/src/cpp/rtps/common/GuidPrefix_t.cpp
+    ${PROJECT_SOURCE_DIR}/src/cpp/rtps/common/SerializedPayload.cpp
     ${PROJECT_SOURCE_DIR}/src/cpp/rtps/common/LocatorWithMask.cpp
```
Comment (MiguelCompany) [inline test/unittest/rtps/reader/CMakeLists.txt:208]: Move SerializedPayload 1 line down

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

**Comment by MiguelCompany** [comment]
@richiprosima Please test_3 discovery-server

**Comment by MiguelCompany** [comment]
Checked locally that discovery server tests pass with this and #4932. Ready to merge.

**Comment by MiguelCompany** [comment]
@richiprosima Please test_3 discovery-server because #4932 has just been merged

**Comment by MiguelCompany** [comment]
@richiprosima Please test_3 discovery-server with updated ci branch

**Comment by MiguelCompany** [comment]
@richiprosima Please test_3 discovery-server because I made a mistake in the jenkins job configuration

**Pull request merged**